samples/functions/overloading_different_return_types.jakt

```diff
@@ -0,0 +1,15 @@
+/// Expect:
+/// - output: "true\ntrue\ntrue\n"
+
+function identity(anon x: i64) throws => x
+function identity(anon x: String) => x
+function identity<T>(anon x: T) => x
```

Comment:
What happens if you declare the generic function before the concrete overloads? Does it still prefer the concrete functions before instantiating the generic?

Reply:
I'm a wee bit tempted to say we should try them in the order you declare them, and the first one that succeeds should win. Might be worth chatting through on discord (if we haven't already)

Reply:
Currently all matching functions get scored by the function described in https://github.com/SerenityOS/jakt/pull/1202/commits/8185aac078223f97c30fa0e401895f74a0ac5ec1 and the highest score wins, so yes, the concrete function gets picked over the generic version regardless of declaration order.

Reply:
Is that implemented via name mangling?

I think using declaration order is not a good idea, declaration order can also be changed by shuffling imports or just cleaning up code to make it more readable, seems awfully fragile

Reply:
What is the use case for allowing such overloads?

Reply:
> Is that implemented via name mangling?

The current implementation does not mangle any names. Consent on discord was that matching the (relevant) c++ rules is worth the tradeoff for increased compatibility instead of name mangling and doing our own thing (for now?)

> What is the use case for allowing such overloads?

One concrete example would be a catch-all method for variants like this:

```
boxed enum Box {
    Void
    Bool(value:bool)
    U8(value: u8)
    U16(value: u16)
    ..
    JaktString(value: String)
    JaktArray(value: Array<Box>)
    JaktDictionary(value: Dictionary<Box, Box>)
    JaktSet(value: Set<Box>)
    JaktOptional(value: Box)
    JaktNone
    Unknown

    function create(value: String) => Box::JaktString(value)
    function create(value: u8) => Box::U8(value)
    function create(values: Array<T>) {
        let jakt_values: Box = []
        for value in values.iterator() {
            jakt_values.push(Box::create(value))
        }
        Box::Array(value: jakt_values)
    }
    function create(value: T) => Box::Unknown
}
```